core/BranchPred.cpp

```diff
@@ -0,0 +1,70 @@
+#include "BranchPred.hpp"
+
+/*
+ * The algorithm used for prediction / update is as follows:
+ * Prediction:
+ *    - look up BHT to determine if the branch is predicted taken or not
+ *    - look up BTB to see if an entry exists for the input fetch pc
+ *       - if present in BTB and predicted taken, BTB entry is used to determine
+ *         prediction branch idx and predictedPC
+ *       - if present in BTB but predicted not taken, BTB entry is used to determine
+ *         prediction branch idx, while predictedPC is the fall through addr
+ *       - if not present in BTB entry, prediction branch idx is the last instr of
+ *         the FetchPacket, while predicted PC is the fall through addr. Also, create
+ *         a new BTB entry 
+ * Update:
+ *    - a valid BTB entry must be present for fetch PC
+ *    - TBD 
+ *
+ */
+
+#define BYTES_PER_INST 4
+
+void SimpleBranchPredictor::updatePredictor(DefaultUpdate & update) {
+
+    sparta_assert(branch_target_buffer_.find(update.FetchPC) != branch_target_buffer_.end());
+    branch_target_buffer_[update.FetchPC].branch_idx = update.branch_idx;
+    if (update.actuallyTaken) {
+        branch_history_table_[update.FetchPC] =
+            (branch_history_table_[update.FetchPC] == 3) ? 3 :
+             branch_history_table_[update.FetchPC] + 1;
+        branch_target_buffer_[update.FetchPC].predictedPC = update.correctedPC;
+    } else {
+        branch_history_table_[update.FetchPC] =
+            (branch_history_table_[update.FetchPC] == 0) ? 0 :
+             branch_history_table_[update.FetchPC] - 1;
+    }
+}
+
+DefaultPrediction SimpleBranchPredictor::getPrediction(const DefaultInput & input) {
+    bool predictTaken;
+    if (branch_history_table_.find(input.FetchPC) != branch_history_table_.end()) {
+        predictTaken = (branch_history_table_[input.FetchPC] > 1);
+    } else {
+        predictTaken = false;
+        // add a new entry to BHT, biased towards not taken
+        branch_history_table_.insert(std::pair<uint64_t, uint8_t>(input.FetchPC, 1));       
```

Comment:
Is there a reason this is initialized to 1 and not 0?
branch_history_table_.insert(std::pair<uint64_t, uint8_t>(input.FetchPC, 0));

Reply:
I believe this is initialized to 1 to indicate a starting point in the state-machine of "weakly not taken" vs. 0 which is "strongly not taken".

It may be worth explicitly stating the state-machine states in the header description for the implemented "SimpleBranchPredictor", just so it is clear what algorithm is implemented.  Or maybe just link to a description of what algorithm is implemented (e.g. wiki).

Reply:
Expanded comment at the beginning of this file as an explanation.